### PR TITLE
doc(common): in-depth guide for `StatusOr`

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #10157

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10555)

### CODE REVIEW

File: google/cloud/doc/error-handling.dox [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+Some functions return a value that already has a mechanism to signal failures.
+For example, some functions return [AsyncStreamingReadWriteRpc<T,U>] (or
+technically `std::unique_ptr<AsyncStreamingReadWriteRpc<T,U>>`), and other
+functions return classes derived from `std::istream`. In such
+cases the library does not wrap the result in a `StatusOr<T>` because the
+returned type already has mechanisms to signal errors.
+
+## Futures
+
+Some functions return a "future" ([future<T>]). This objects represent a value
```
Comment (devbww) [inline google/cloud/doc/error-handling.dox:39]: s/This/These/

File: google/cloud/samples/samples.cc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  std::cout << "\nRunning StatusOrExceptions() example [2]" << std::endl;
+  StatusOrExceptions({"projects/my-project-id"});
+}
+
+}  // namespace
+
+int main(int argc, char* argv[]) {  // NOLINT(bugprone-exception-escape)
+  google::cloud::testing_util::Example example({
+      {"status-or-usage", StatusOrUsage},
+      {"status-or-exceptions", StatusOrExceptions},
```
Comment (devbww) [inline google/cloud/samples/samples.cc:90]: Do we anticipate that one file, `samples.cc`, will be sufficient for all samples?  Or might we go with `status_samples.cc` in this case.  Or (and I imagine your opinion), should we just wait and see?

**Review by devbww — APPROVED** [review]

File: google/cloud/doc/error-handling.dox [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+[future<T>]: @ref google::cloud::future
+[RuntimeStatusError]: @ref google::cloud::RuntimeStatusError
+[error_info()]: @ref google::cloud::Status::error_info()
+[input iterators]: https://en.cppreference.com/w/cpp/named_req/InputIterator
+[iostream operator<<]: @ref google::cloud::operator<<(std::ostream&,google::cloud::Status const&)
+
+## Overview
+
+In general, the `google-cloud-cpp` libraries return a [StatusOr<T>] if a
+function many fail and needs to signal an error. `StatusOr<T>` is an "outcome",
```
Comment (tmatsuo) [inline google/cloud/doc/error-handling.dox:17]: s/many/may/

**Review by tmatsuo — APPROVED** [review]

File: google/cloud/doc/error-handling.dox (hunk 3 above)
Comment (coryan) [reply google/cloud/doc/error-handling.dox]: Fixed

File: google/cloud/samples/samples.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/samples/samples.cc:90]: Tempting... But I am going to YAGNI this (aka "wait and see")

File: google/cloud/doc/error-handling.dox (hunk 1 above)
Comment (coryan) [reply google/cloud/doc/error-handling.dox]: Fixed

**Review by devbww — APPROVED** [review]

**Pull request merged**